tests/test_warnings.cpp

```diff
@@ -0,0 +1,79 @@
+/*
+    tests/test_warnings.cpp -- usage of warnings::warn() and warnings categories.
+
+    Copyright (c) 2024 Jan Iwaszkiewicz <jiwaszkiewicz6@gmail.com>
+
+    All rights reserved. Use of this source code is governed by a
+    BSD-style license that can be found in the LICENSE file.
+*/
+
+#include <pybind11/warnings.h>
+
+#include "pybind11_tests.h"
+
+#include <utility>
+
+namespace warning_helpers {
+void warn_function(py::module &m, const char *name, py::handle category, const char *message) {
+    m.def(name, [category, message]() { py::warnings::warn(message, category); });
+}
+} // namespace warning_helpers
+
+class CustomWarning {};
+
+TEST_SUBMODULE(warnings_, m) {
+
+    // Test warning mechanism base
+    m.def("raise_and_return", []() {
+        std::string message = "Warning was raised!";
+        py::warnings::warn(message.c_str(), PyExc_Warning);
+        return 21;
+    });
+
+    m.def("raise_default", []() { py::warnings::warn("RuntimeWarning is raised!"); });
+
+    m.def("raise_from_cpython",
+          []() { py::warnings::warn("UnicodeWarning is raised!", PyExc_UnicodeWarning); });
+
+    m.def("raise_and_fail",
+          []() { py::warnings::warn("RuntimeError should be raised!", PyExc_Exception); });
```

Comment:
This confused me:

`"RuntimeError should be raised!"`

Suggestion to change to:

`"Invalid category"`

I'd also change `raise_and_fail` to `raise_invalid_category`

Actually: is "raise" a good term to use here (all test functions)?

Because there is no "raise" in `warnings.warn()`.

Maybe `warn_with_invalid_category` etc? So that it is more clear to someone who gets here from a different context (e.g. refactoring or cleanup) what's happening.

Reply:
I see the motivation, I hope now the same are more self-explainatory/clear.